File: android/src/main/kotlin/com/mparticle/mparticle_flutter_sdk/InitializeOptionsParser.kt
package com.mparticle.mparticle_flutter_sdk

import com.mparticle.MParticle
import java.net.URI

internal sealed class CustomBaseUrlValidation {
  object Absent : CustomBaseUrlValidation()
  data class Valid(val url: String) : CustomBaseUrlValidation()
  data class Invalid(val reason: String) : CustomBaseUrlValidation()
}

internal object InitializeOptionsParser {
  private const val MAX_BOOTSTRAP_IDENTITIES = 10
  private const val MAX_IDENTITY_VALUE_LENGTH = 256

  fun validateBootstrapIdentities(identityMap: Map<String, String>): String? {
    if (identityMap.size > MAX_BOOTSTRAP_IDENTITIES) {
      return "bootstrapIdentityRequest exceeds maximum identity count."
    }
    val parsedKeys = mutableSetOf<Int>()
    for ((key, value) in identityMap) {
      val intKey = key.toIntOrNull()
      if (intKey == null) {
        return "bootstrapIdentityRequest contains invalid identity keys."
      }
      if (!parsedKeys.add(intKey)) {
        return "bootstrapIdentityRequest contains duplicate identity keys."
      }
      if (value.isBlank()) {
        return "bootstrapIdentityRequest contains empty values."
      }
      if (value.length > MAX_IDENTITY_VALUE_LENGTH) {
        return "bootstrapIdentityRequest value exceeds maximum length."
      }
    }
    return null
  }

  fun parseLogLevel(index: Int): MParticle.LogLevel {
    return when (index) {
      0 -> MParticle.LogLevel.NONE
      1 -> MParticle.LogLevel.ERROR
      2 -> MParticle.LogLevel.WARNING
      3 -> MParticle.LogLevel.INFO
      4 -> MParticle.LogLevel.DEBUG
      5 -> MParticle.LogLevel.VERBOSE
      else -> MParticle.LogLevel.WARNING
    }
  }

  fun parseEnvironment(index: Int): MParticle.Environment {
    return when (index) {
      1 -> MParticle.Environment.Development
      2 -> MParticle.Environment.Production
      else -> MParticle.Environment.AutoDetect
    }
  }

  fun validateCustomBaseUrl(raw: String?): CustomBaseUrlValidation {
    if (raw.isNullOrBlank()) {
      return CustomBaseUrlValidation.Absent
    }
    val trimmed = raw.trim()
    if (!trimmed.startsWith("https://", ignoreCase = true)) {
      return CustomBaseUrlValidation.Invalid("customBaseUrl must use https://")
    }
    val host = try {
      URI(trimmed).host
    } catch (_: Exception) {
      null
    }
    if (host.isNullOrEmpty()) {
      return CustomBaseUrlValidation.Invalid("customBaseUrl is not a valid https URL")
    }
    return CustomBaseUrlValidation.Valid(trimmed)
  }
}

File: ios/mparticle_flutter_sdk/InitializeOptionsParserPackage/Sources/InitializeOptionsParser/InitializeOptionsParser.swift
/// Maps Dart [MparticleOptions] wire indices to mParticle Apple SDK raw values.
///
/// Kept UIKit-free so unit tests run on macOS without linking mParticle-Apple-SDK.
import Foundation

public enum InitializeOptionsParser {
  private static let maxBootstrapIdentities = 10
  private static let maxIdentityValueLength = 256

  /// Returns an error message when bootstrap identities violate Dart wire limits.
  public static func validateBootstrapIdentities(_ identities: [String: String]) -> String? {
    if identities.count > maxBootstrapIdentities {
      return "bootstrapIdentityRequest exceeds maximum identity count."
    }
    var parsedKeys = Set<Int>()
    for (key, value) in identities {
      guard let intKey = Int(key) else {
        return "bootstrapIdentityRequest contains invalid identity keys."
      }
      if parsedKeys.contains(intKey) {
        return "bootstrapIdentityRequest contains duplicate identity keys."
      }
      parsedKeys.insert(intKey)
      if value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
        return "bootstrapIdentityRequest contains empty values."
      }
      if value.count > maxIdentityValueLength {
        return "bootstrapIdentityRequest value exceeds maximum length."
      }
    }
    return nil
  }

  /// Maps Dart log level index (0–5) to `MPILogLevel` raw value.
  ///
  /// Apple SDK has no INFO level; Dart index 3 (info) maps to Debug (3).
  public static func parseLogLevelRawValue(_ index: Int) -> UInt {
    switch index {
    case 0:
      return 0 // MPILogLevelNone
    case 1:
      return 1 // MPILogLevelError
    case 2:
      return 2 // MPILogLevelWarning
    case 3, 4:
      return 3 // MPILogLevelDebug
    case 5:
      return 4 // MPILogLevelVerbose
    default:
      return 2 // MPILogLevelWarning
    }
  }

  /// Maps Dart environment index to `MPEnvironment` raw value.
  public static func parseEnvironmentRawValue(_ index: Int) -> UInt {
    switch index {
    case 1:
      return 1 // Development
    case 2:
      return 2 // Production
    default:
      return 0 // AutoDetect
    }
  }
}
